src/query/service/src/pipelines/processors/transforms/hash_join/hash_join_build_state.rs

```diff
@@ -181,14 +182,17 @@ impl HashJoinBuildState {
     /// Add input `DataBlock` to `hash_join_state.row_space`.
     pub fn build(&self, input: DataBlock) -> Result<()> {
         let mut buffer = self.hash_join_state.row_space.buffer.write();
+
         let input_rows = input.num_rows();
-        buffer.push(input);
+        buffer.push(input.clone());
         let old_size = self
             .hash_join_state
             .row_space
             .buffer_row_size
             .fetch_add(input_rows, Ordering::Relaxed);
 
+        self.merge_into_try_build_block_info_index(input.clone(), old_size);
```

Comment:
We can execute `merge_into_try_build_block_info_index` before `buffer.push(input)` and clone when necessary in `merge_into_try_build_block_info_index`, so that we can reduce one clone operation.

Reply:
well we can remove clone, but we need the old size, so let's put it here.

Reply:
How about putting `merge_into_try_build_block_info_index` and `old_size` together in front of `buffer.push(input)`.